### PR TITLE
Support attaching uprobes to targets inside containers

### PR DESCRIPTION
**Pull request by anfredette**
This pr adds support for attaching uprobes to targets inside containers.

`namespace` has been renamed to `namespace_pid` in bpfd proto and changed to an int32. When the user provides a `namespace_pid`, bpfd will attempt to attach the uprobe inside the container with the given PID.

We use a new executable called `bpfd-ns` which can be called to attach uprobes and uretprobes inside containers.  `bpfd-ns` is spawned from bpfd when a `namespace_pid` parameter is provided when loading a uprobe. `bpfd-ns` changes to the mount namespace of the container, does the attach, and then changes back to the original mount namespace.

### CODE REVIEW

**Comment by anfredette** [comment]
This is working and ready for review -- I'd be glad to get some feedback (even if it's general).

I marked it as a draft because I want to add an integration test before merging it.

File: tests/integration-test/src/tests/utils.rs
Comment (astoycos) [inline tests/integration-test/src/tests/utils.rs:961]: Do we want to be adding a dependency on docker in our core integration tests? Could we just do a 

```
[astoycos@nfvsdn-02-oot bpfman]$ unshare -Urfp --mount-proc

[astoycos@nfvsdn-02-oot bpfman]$ lsns | grep bash
4026533388 pid         1 3416159 astoycos └─-bash

[astoycos@nfvsdn-02-oot bpfman]$ sudo nsenter -t 3416159 -a ls
```

To basically mimic the same thing without docker?

File: Cargo.lock
Comment (astoycos) [inline Cargo.lock:425]:
```suggestion
version = "0.3.1"
```

**Review by astoycos — CHANGES_REQUESTED** [review]
Looking good a bunch of questions really! Also I just am looking to make sure this is generic enough to also use for exec-ing into network namespaces as well for TC and XDP. 

Additionally with the move to damonless, do we for-see all of this functionality moving back into core rather than it's own binary?

**Comment by anfredette** [comment]
> Looking good a bunch of questions really! Also I just am looking to make sure this is generic enough to also use for exec-ing into network namespaces as well for TC and XDP. 

Additionally with the move to damonless, do we for-see all of this functionality moving back into core rather than it's own binary?   

I belive it can be extended for other types as well, but I wasn't planning to do that unless there's a specific demand due to our change in priorities.

We should be able to move this functionality into the core code if it becomes single threaded.

File: bpfman-ns/src/main.rs
Comment (anfredette) [reply bpfman-ns/src/main.rs:22]: pid is optional.  It means "only execute uprobe if the function is called by the given pid".  namespace_pid is optional for the user api, but is required for this executable.  I think that when we extend this to other program types, program_pin_path and namespace_pid (or whatever we call it per comment below) would be common required parameters.

Regarding your second question, "namespace_pid" may not be a good name for the parameter.  This feature was originally billed as attaching uprobes in namespaces, so we started with "namespace" and I stuck with the theme. However, this is really the host PID for the container in which we want to attach the eBPF program.  Namespace isn't required except possibly as part of a selector in k8s.  My inclination is to change this parameter name to "container_pid" unless someone has better idea.

File: bpfman-ns/src/main.rs
Comment (anfredette) [reply bpfman-ns/src/main.rs:36]: Good question.  It's optional for the user api, but I made it required for the api of this executable.  I set it to zero if the user didn't give an offset.  I'll update the comment to make this more clear.

File: bpfman-ns/src/main.rs
Comment (anfredette) [reply bpfman-ns/src/main.rs:114]: It was useful for debugging.  And, if this fails in real life, I think we'll want to know why.  I'll see what it looks like if I do a wrapper function like you suggest below.

File: bpfman-ns/src/main.rs
Comment (anfredette) [reply bpfman-ns/src/main.rs:124]: I moved the setns out and cleaned up some debug logs.

File: bpfman/src/bpf.rs
Comment (anfredette) [reply bpfman/src/bpf.rs:618]: I changed it a little to use `if let`, but if I try to do everything inside the `if let`, rust complains about the lifetime of `pid`.  So, I addressed that with the initial `let mut pid`.  If I try to do the `prog_args.extend` inside the `if let` rust complains that pid is assigned to but never read.  That doesn't make sense to me, but ¯\_(ツ)_/¯.  I don't know if there is a better way.

File: bpfman/src/bpf.rs
Comment (anfredette) [reply bpfman/src/bpf.rs:570]: That's better.

File: tests/integration-test/src/tests/utils.rs
Comment (anfredette) [reply tests/integration-test/src/tests/utils.rs:961]: To be honest, I don't know.  It would be simpler and probably less overhead to not use it.  I asked in one of the team meetings, and people seemed to be okay with using Docker.  I saw https://www.redhat.com/sysadmin/pid-namespace where they discussed using the command you suggest to "build containers by hand".  Is this exactly equivalent to a Docker container and/or the ones used in k8s, or is it safer to just using Docker?  Personally, I'd have to do some more research to be sure.

**Comment by anfredette** [comment]
@astoycos, Thanks for the review.  I think I've addressed most of your comments, but I didn't do them all exactly as suggested, so please take a look.

However, there are two things that I'd like to get consensus on before moving forward:

1. I think the parameter `namespace_pid` is confusing and should be changed.  I'm thinking `container_pid`.  We originally talked about attaching these programs in "namespaces", but it's not really a "namespace" in the k8s sense but rather a container (which does use linux namespaces++ and may be in a different k8s namespace from the user, but may not).  This change will ripple through the code, so I'd like to get some consensus before I do it.

1. Should we use Docker to create containers for integration test, or is it better to create our own "containers" using the Linux command line?

File: bpfman/src/bpf.rs
Comment (dave-tucker) [inline bpfman/src/bpf.rs:603]:
```suggestion
                        let mut prog_args = vec![
                            "uprobe".to_string(),
                            "--program-pin-path".to_string(),
                            program_pin_path,
                            "--offset".to_string(),
                            offset,
                            "--target".to_string(),
                            program.target,
                            "--namespace-pid".to_string(),
                            namespace_pid,
                        ];
```

File: bpfman/src/bpf.rs
Comment (dave-tucker) [inline bpfman/src/bpf.rs:608]:
```suggestion
                        if let Some(fn_name) = program.fn_name {
                            prog_args.extend(["--fn-name".to_string(), fn_name])
                        }
```

File: bpfman/src/bpf.rs
Comment (dave-tucker) [inline bpfman/src/bpf.rs:612]:
```suggestion
                        if program.retprobe {
                            prog_args.push("--retprobe".to_string());
                        }
```

File: bpfman/src/bpf.rs
Comment (dave-tucker) [inline bpfman/src/bpf.rs:620]:
```suggestion
                        if let Some(pid) = program.pid {
                            prog_args.extend(["--pid".to_string(), pid.to_string()])
                        }
```

**Review by dave-tucker — CHANGES_REQUESTED** [review]

**Review by dave-tucker — COMMENTED** [review]
Some nits around the way you build the args to the bpfman-ns binary but otherwise LGTM once Andrew's comments are addressed.

**Comment by dave-tucker** [comment]
> @astoycos, Thanks for the review. I think I've addressed most of your comments, but I didn't do them all exactly as suggested, so please take a look.
> 
> However, there are two things that I'd like to get consensus on before moving forward:
> 
> 1. I think the parameter `namespace_pid` is confusing and should be changed.  I'm thinking `container_pid`.  We originally talked about attaching these programs in "namespaces", but it's not really a "namespace" in the k8s sense but rather a container (which does use linux namespaces++ and may be in a different k8s namespace from the user, but may not).  I'd also change the executable from `bpfman-ns` to something like `bpfman-container`.  This change will ripple through the code, so I'd like to get some consensus before I do it.

`container_pid` makes sense. Stay with `bpfman-ns` - it's not user-facing so we can change it whenever.
Once this has merged we can look at folding this into a dedicated (hidden) command like `bpfman attach-in-container` or something similar.

> 2. Should we use Docker to create containers for integration test, or is it better to create our own "containers" using the Linux command line?

No. We should use https://github.com/testcontainers/testcontainers-rs
IIRC testcontainers had some support for podman - and we can document both in the integration-test docs.
It will be a lot easier than manually creating fake containers ourselves.

**Comment by astoycos** [comment]
So talking to @anfredette offline, I wrote previously 

> From an external(API) POV I think the container ID/Name is what should be provided by the user. So then a user can easily express that the uprobe should be attached to the process in the specified container and bpfman internally can just map the container name to the actual container PID. Having a user have to specify containerPID seems ultra low level.

Due to the fact that using a container ID/name would require us to interface more closely with a container runtime (podman, Docker) I'm actually alright with just using container_pid as the API for now.

File: proto/bpfman.proto
Comment (anfredette) [reply proto/bpfman.proto:113]: Good point.  I don't mind changing this, but it may not be necessary yet.  First, the namespace option isn't supported yet.  There's no way to use it from kubernetes, and if someone tries it from bpfctl or bpfman, they'll get an error.  Second, we're going to be making major changes in 0.4.0, so I don't know if we need to worry about backward compatibiilty yet. Let's discuss during the meeting tomorrow.

File: bpfman/src/bpf.rs
Comment (anfredette) [reply bpfman/src/bpf.rs]: Done

File: bpfman/src/bpf.rs
Comment (anfredette) [reply bpfman/src/bpf.rs:615]: Done.

File: bpfman/src/bpf.rs
Comment (anfredette) [reply bpfman/src/bpf.rs:607]: Done

**Review by astoycos — APPROVED** [review]
LGTM Thanks @anfredette 

Please remember to make any follow up issues we discussed

**Review by dave-tucker — APPROVED** [review]
LGTM 🌴

**Pull request merged**
